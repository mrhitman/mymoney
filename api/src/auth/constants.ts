import { config } from 'dotenv';

config();

export const jwtConstants = {
  secret: process.env.JWT_SECRET,
  expiresIn: 5 * 60,
};
