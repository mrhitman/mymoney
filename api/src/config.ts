import * as dotenv from 'dotenv';
import { get, isNull, isUndefined } from 'lodash';

dotenv.config({ path: process.env.PWD + '/.env' });

export function getEnvVariable(name: string, defaultValue?: string) {
  const value = get(process.env, name, defaultValue);

  if (isNull(value) || isUndefined(value)) {
    throw new Error(`Required env variable:${name} isn't set`);
  }

  return value;
}

export const config = {
  app: {
    maxQueryComplexity: parseInt(getEnvVariable('MAX_QUERY_COMPLEXITY', '32')),
    jwt: {
      secret: getEnvVariable('JWT_SECRET'),
      expiresIn: getEnvVariable('JWT_EXPIRES_IN', '10m'),
      ignoreExpiration: Boolean(+getEnvVariable('JWT_IGNORE_EXPIRATION', '0')),
    },
    port: parseInt(getEnvVariable('PORT', '3000'), 10),
    host: getEnvVariable('HOST'),
  },
  mail: {
    type: 'OAuth2',
    user: getEnvVariable('MAIL_USER'),
    clientId: getEnvVariable('MAIL_CLIENT_ID'),
    clientSecret: getEnvVariable('MAIL_SECRET'),
    accessToken: getEnvVariable('MAIL_TOKEN'),
    refreshToken: getEnvVariable('MAIL_REFRESH_TOKEN'),
  },
  db: {
    client: 'postgres',
    connection: {
      connectionString: getEnvVariable('DATABASE_URL'),
      ssl: Boolean(+getEnvVariable('DATABASE_SSL', '1'))
        ? { rejectUnauthorized: false }
        : false,
    },
    timezone: 'UTC',
    charset: 'utf8',
    debug: Boolean(+getEnvVariable('KNEX_DEBUG', '0')),
    pool: {
      min: 2,
      max: 16,
    },
  },
  redis: {
    url: getEnvVariable('REDIS_URL'),
  },
  fixer: {
    apiKey: getEnvVariable('FIXER_API_KEY'),
  },
};

export type Config = typeof config;
