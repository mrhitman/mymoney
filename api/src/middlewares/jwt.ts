import jwt from 'koa-jwt';

export default jwt({ secret: process.env.SALT, key: 'jwtdata' });
