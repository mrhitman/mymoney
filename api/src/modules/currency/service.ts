import axios from 'axios';
import xml2js from 'xml2js';
import redis from '../../services/redis';

class CurrencyProvider {
  async latest(base?: string) {
    const client = redis;
    const rates = await client.get('rates');
    if (rates) {
      return JSON.parse(rates);
    }
    const response = await fixer.latest({ base });
    await client.set('rates', JSON.stringify(response), 'EX', 3600);
    return response;
  }

  // ISO_4217 standart
  async getIsoInfo() {
    const url = 'https://www.currency-iso.org/dam/downloads/lists/list_one.xml';
    const body = await axios.get(url);
    return new Promise((resolve, reject) => {
      xml2js.parseString(body.data, { explicitArray: false }, (err, result) => {
        if (err) {
          return reject(err);
        }
        resolve(result);
      });
    });
  }
}
