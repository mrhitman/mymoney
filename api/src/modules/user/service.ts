import { hash } from 'bcrypt';
import User from 'src/models/user';
import { BadRequest, NotFound } from 'ts-httpexceptions';
import CreateUserDto from './dto/CreateUserDto';
import UpdateUserDto from './dto/UpdateUserDto';
import { userInfo } from 'os';

export class UserProvider {
  public async create(dto: CreateUserDto) {
    if (await User.query().findOne({ email: dto.email })) {
      throw new BadRequest('User already exists');
    }

    return User.query().insert({
      email: dto.email,
      first_name: dto.first_name,
      middle_name: dto.middle_name,
      last_name: dto.last_name,
      password: await hash(dto.password, 10),
      last_sync: new Date(),
    });
  }

  public async update(dto: UpdateUserDto) {
    const user = await User.query().findOne({ email: dto.email });

    if (!user) {
      throw new NotFound('No such user found');
    }

    await user.$query().update({ last_sync: new Date() });
    return user;
  }

  public async get() {
    return User.query();
  }
}

export default UserProvider;
