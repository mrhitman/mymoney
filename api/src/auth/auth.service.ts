import { BadRequestException, Injectable } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import bcrypt from 'bcrypt';
import { omit } from 'lodash';
import RefreshToken from 'src/database/models/refresh-token.model';
import User from 'src/database/models/user.model';
import { UsersService } from 'src/users/users.service';
import { v4 as uuid } from 'uuid';

@Injectable()
export class AuthService {
  constructor(
    protected readonly usersService: UsersService,
    protected readonly jwtService: JwtService,
  ) { }

  public async validateUser(email: string, password: string) {
    const user = await this.usersService.findByEmail(email);
    const hash = await bcrypt.hash(password, 10);

    if (user && bcrypt.compare(password, hash)) {
      return omit(user, ['password']);
    }
  }

  public async login(user: User, token?: RefreshToken) {
    const payload = { id: user.id };
    const refreshToken = token ? token.token : uuid();

    if (!token) {
      const refreshToken = uuid();
      await RefreshToken.query().delete().where({ userId: user.id });
      await RefreshToken.query().insert({ userId: user.id, token: refreshToken });
    }

    return {
      accessToken: this.jwtService.sign(payload),
      refreshToken,
    };
  }

  public async logout(user: User) {
    await RefreshToken.query().delete().where({ userId: user.id });
  }

  public async refresh(token: string) {
    const refreshToken = await RefreshToken.query().where({ token }).first();

    if (!refreshToken) {
      throw new BadRequestException('Invalid refresh token');
    }

    const user = await this.usersService.findById(refreshToken.userId);
    return this.login(user, refreshToken);
  }

  public async getUser(id: number) {
    const user = await this.usersService.findById(id);

    return omit(user, ['password']);
  }
}
