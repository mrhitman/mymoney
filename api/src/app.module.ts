import { Module } from '@nestjs/common';
import { UsersModule } from './users/users.module';
import { WalletsModule } from './wallets/wallets.module';
import { CategoriesModule } from './categories/categories.module';
import { BudgetsModule } from './budgets/budgets.module';
import { GoalsModule } from './goals/goals.module';
import { TransactionsModule } from './transactions/transactions.module';
import { AppController } from './app/app.controller';
import { CurrenciesModule } from './currencies/currencies.module';
import { Fixer } from './fixer';
import { AuthModule } from './auth/auth.module';

@Module({
  imports: [
    UsersModule,
    WalletsModule,
    CategoriesModule,
    BudgetsModule,
    GoalsModule,
    TransactionsModule,
    CurrenciesModule,
    AuthModule,
  ],
  controllers: [AppController],
  providers: [Fixer],
})
export class AppModule {}
