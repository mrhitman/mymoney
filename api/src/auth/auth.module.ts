import { Module } from '@nestjs/common';
import { JwtModule } from '@nestjs/jwt';
import { PassportModule } from '@nestjs/passport';
import { UsersModule } from 'src/users/users.module';
import { UsersService } from 'src/users/users.service';
import { AuthService } from './auth.service';
import { config } from 'src/config';
import { JwtStrategy } from './strategies/jwt.strategy';
import { LocalStrategy } from './strategies/local.strategy';

const Jwt = JwtModule.register({
  secret: config.app.jwt.secret,
  signOptions: { expiresIn: config.app.jwt.expiresIn },
});
@Module({
  imports: [UsersModule, PassportModule, Jwt],
  providers: [AuthService, UsersService, LocalStrategy, JwtStrategy],
  exports: [AuthService, Jwt],
})
export class AuthModule {}
