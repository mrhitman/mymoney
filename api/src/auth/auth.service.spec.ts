import { JwtModule } from '@nestjs/jwt';
import { Test, TestingModule } from '@nestjs/testing';
import { Chance } from 'chance';
import { config } from 'src/config';
import { DatabaseModule } from 'src/database/database.module';
import Category from 'src/database/models/category.model';
import UserCategory from 'src/database/models/user-category.model';
import User from 'src/database/models/user.model';
import { UsersService } from 'src/users/users.service';
import { AuthService } from './auth.service';

describe('AuthService', () => {
  let service: AuthService;

  beforeEach(async () => {
    jest.clearAllMocks();
    const module: TestingModule = await Test.createTestingModule({
      imports: [
        DatabaseModule,
        JwtModule.register({
          secret: 'test_secret',
          signOptions: { expiresIn: config.app.jwt.expiresIn },
        }),
      ],
      providers: [AuthService, UsersService],
    }).compile();

    service = module.get<AuthService>(AuthService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe(' register User', () => {
    const chance = new Chance(42);
    const userData = {
      firstName: chance.name(),
      email: chance.email(),
      password: chance.word({ length: 10 }),
      imageUrl: '',
      additional: {},
    };
    beforeEach(() => {
      (<jest.Mock>User.query).mockImplementation(() => ({
        findOne: jest.fn().mockReturnValue(null),
        insert: jest.fn().mockReturnValue({ id: 1, ...userData }),
      }));
      (<jest.Mock>UserCategory.query).mockImplementation(() => ({
        where: jest.fn().mockReturnThis(),
        insert: jest
          .fn()
          .mockReturnValue([
            { id: 1, name: 'test_category', type: 'income', userId: 1 },
          ]),
        count: jest.fn().mockReturnValue([{ count: 0 }]),
      }));
      (<jest.Mock>Category.query).mockImplementation(() => [
        {
          id: 1,
          name: 'test_category',
          type: 'income',
          codes: [],
          icon: {},
          isFixed: false,
        },
      ]);
    });

    it(' successfully', async () => {
      const stab = jest.fn();
      (<jest.Mock>UserCategory.query).mockImplementation(() => ({
        where: jest.fn().mockReturnThis(),
        insert: stab.mockReturnValue([
          { id: 1, name: 'test_category', type: 'income', userId: 1 },
        ]),
        count: jest.fn().mockReturnValue([{ count: 0 }]),
      }));
      const user = await service.register(userData);
      expect(user).toBeDefined();
      expect(user.firstName).toBe(userData.firstName);
      expect(user.email).toBe(userData.email);
      expect(user).toMatchSnapshot();
      expect(stab).toBeCalledWith([
        {
          categoryId: 1,
          codes: [],
          icon: {},
          id: 'RANDOM_GENERETED_UUID_V4:40',
          isFixed: false,
          name: 'test_category',
          type: 'income',
          userId: 1,
        },
      ]);
    });

    it(' failed, user already exists', async () => {
      (<jest.Mock>User.query).mockImplementation(() => ({
        findOne: jest.fn().mockReturnValue({ id: 1, ...userData }),
      }));
      await expect(
        service.register(userData),
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });
});
