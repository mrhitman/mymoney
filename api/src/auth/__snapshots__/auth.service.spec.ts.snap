// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthService  register User  failed, user already exists 1`] = `"Such email is busy"`;

exports[`AuthService  register User  successfully 1`] = `
Object {
  "additional": Object {},
  "email": "zin@fa.am",
  "firstName": "Seth Tran",
  "id": 1,
  "imageUrl": "",
  "password": "kodtacwuwb",
}
`;
